### PR TITLE
Add dedicated get_page perf test

### PR DESCRIPTION
**Pull request by bojanserafimov**
Resolves https://github.com/zenithdb/zenith/issues/1380

This test:
1. Starts up the zenith env, and runs a workload to fill up the pageserver with some wal
2. Checkpoints the in-mem layer
3. Runs individual get_page queries at latest lsn by calling the pagestream API directly from a rust program, psbench
4. Reports runtimes back to the python benchmarker tool

Results from running:
```
test_get_page.num_pages: 9741
test_get_page.num_wal_entries: 52611
test_get_page.total_wal_size: 71106851 bytes
test_get_page.fastest: 21 microseconds
test_get_page.median: 49 microseconds
test_get_page.p99: 206 microseconds
test_get_page.slowest: 6179 microseconds
```

See [here](https://github.com/neondatabase/neon/issues/1380#issuecomment-1089575807) for more experimental results.

TODO after merge:
- consider adding more capabilities to psbench, for different kinds of tests (for example, querying near-latest LSN)
- test using different setup workloads (for [example](https://github.com/neondatabase/neon/issues/1380#issuecomment-1089575807))
- run this test remotely. Maybe deploy a pageserver on one ec2 machine, and all other components, including wal_metadata logger on another machine, and run the test from the latter. Or maybe precompute wal_metadata, restart the pageserver and fill it up with exactly the same wal. Idk yet.

### CODE REVIEW

**Comment by hlinnaka** [comment]
I like the approach!

File: pageserver/src/bin/psbench.rs [hunk 1]
```diff
@@ -0,0 +1,218 @@
+//! Pageserver benchmark tool
+//!
+//! Usually it's easier to write python perf tests, but here the performance
+//! of the tester matters, and the pagestream API is easier to call from rust.
```
Comment (hlinnaka) [inline pageserver/src/bin/psbench.rs:4]: todo: brief description of what the tool does. Something like
```
//! psbench replays a stream of GetPage requests, previously recorded by [HOW?]
```

File: pageserver/src/bin/psbench.rs [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+            .port(15000)
+            .dbname("postgres")
+            .user("zenith_admin")
+            .connect_raw(&mut stream, tokio_postgres::NoTls)
+            .await?;
+
+        // Enter pagestream protocol
+        let init_query = format!("pagestream {} {}", tenant_hex, timeline);
+        tokio::select! {
+            _ = conn => panic!("AAAA"),
```
Comment (hlinnaka) [inline pageserver/src/bin/psbench.rs:50]: AAAA :-)

**Review by hlinnaka — COMMENTED** [review]
Hmm, so the WAL metadata contains a trace of all pages modified by the incoming WAL. And then psbench performs a GetPage request for each of those pages, in the order they were modified in the WAL stream.

I think the idea of a "WAL metadata", containing a stream of requests to replay, makes sense. I'm not sure this is a good way to generate the trace though. How about recording the GetPage requests from a real workload, instead? Or you could call the `collect_keyspace` function to get the whole keyspace, and generate a list of requests from that.

So +1 on the general idea of a storing a "trace" of GetPage requests in a file, and having a tool to replay it. Maybe the trace of requests could be a json file so that you could easily edit it with a text editor or some other script.

**Comment by bojanserafimov** [comment]
> Hmm, so the WAL metadata contains a trace of all pages modified by the incoming WAL. And then psbench performs a GetPage request for each of those pages, in the order they were modified in the WAL stream.
> 
> I think the idea of a "WAL metadata", containing a stream of requests to replay, makes sense. I'm not sure this is a good way to generate the trace though. How about recording the GetPage requests from a real workload, instead? Or you could call the `collect_keyspace` function to get the whole keyspace, and generate a list of requests from that.
> 
> So +1 on the general idea of a storing a "trace" of GetPage requests in a file, and having a tool to replay it. Maybe the trace of requests could be a json file so that you could easily edit it with a text editor or some other script.

Agree, read traces would be very useful. Being able to replay wal would also be useful. But it expands the scope of this PR too much. I rewrote the file docstring to communicate that I want the tool to move in this direction.

My first test is very limited, and as you noted it doesn't even make much use of the wal metadata. It's not a good test to serve as a baseline for performance improvements (because it's read pattern is too synthetic), but I've found it very useful for learning how the system behaves.

**Comment by aome510** [comment]
@bojanserafimov updated the PR description ("addressing" -> "resolves"), so Github knows it is linked to an issue, hence easier to track

**Comment by bojanserafimov** [comment]
Converted to draft since I figured out how to parse wal files and avoid all the wal metadata hacks.

**Comment by bojanserafimov** [comment]
Unassigning myself. Someone take over if this is still a good idea to do

**Pull request closed without merging**